### PR TITLE
Pop observables and coeffs that have been grouped already

### LINKED ISSUE #1389: Hamiltonians created via addition produce incorrect results with `ExpvalCost(optimize=True)`
Good afternoon, 
I was having problem with vqe when I set the flag: optmize = True . I noticed that there were errors when entering Hamiltonians of the form Z1@Z2 + Z2@Z1. We know it is the same as 2.Z1.Z2 but the optimizer seems to have some kind of problem with this kind of Hamiltonians (vqe with optimize = False works correctly in this situation).
Although by preprocessing the Hamiltonian we avoid the problem, I was writing to take the problem into account.
Greetings!

### PR DESCRIPTION
**Pull request by antalszava**
**Context**

The `ExpvalCost` class uses the grouping module when the `optimize=True` argument is set. Using this option, the terms in the Hamiltonian (that are Pauli words) are grouped based on a relation selected (e.g., qubit-wise commutativity) and then their corresponding coefficients are also grouped accordingly.

That is, grouping involves first the actual grouping of Hamiltonian terms and then grouping the coefficients of the terms.

When there are multiple equivalent terms in the Hamiltonian, with the current logic, the coefficient of the first term that matches a Pauli word will be used when grouping the coefficients. This is so because the grouping module assumes that the terms in the Hamiltonian are unique.

Therefore, equal terms of the Hamiltonian might get the same coefficient incorrectly which will result in incorrect results for the expectation value of the entire Hamiltonian.
```python
import pennylane as qml

a = qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[4])
b = qml.PauliZ(wires=[4]) @ qml.PauliZ(wires=[0])

coeffs, observables = [0.5, 1], [a, b]
obs_groupings, coeffs_groupings = qml.grouping.group_observables(observables, coeffs)
print(coeffs_groupings)
```
```
[[0.5, 0.5]]   # <---- second coeff should be 1
```
*Note*: the logic of `ExpvalCost` does not hinder such a case from being valid and it doesn't explicitly perform the simplification of the Hamiltonian because that might incur considerable overhead for larger Hamiltonians. It is advisable for users to simplify their Hamiltonians in advance.

**Changes**

Updates the grouping mechanism to remove already processed observables and coefficients.

**Benefits**

The grouping module can function even if Hamiltonian terms are not unique.

**Related issues**

Closes #1389

*Further note*: although it should be fair to assume that the terms are unique in the Hamiltonian, making sure that grouping happens even with terms appearing multiple times can ensure that there is no unexpected behaviour. Users should be encouraged to simplify their Hamiltonians at all times.

### CODE REVIEW

File: pennylane/grouping/group_observables.py [hunk 1]
```diff
@@ -236,11 +239,16 @@ def group_observables(observables, coefficients=None, grouping_type="qwc", metho
 
     partitioned_coeffs = [[0] * len(g) for g in partitioned_paulis]
 
+    observables = copy(observables)
```
Comment (antalszava) [inline pennylane/grouping/group_observables.py:242]: Main update starts here

**Review by trbromley — APPROVED** [review]
Thanks @antalszava! Looks good!

You mention it being costly to simplify the Hamiltonian, but I'd be interested to see if it's just as easy to remove repeated observables in the Hamiltonian versus what we do here with the copy. (this is more of a curiosity, we don't need to check! This can all be solved with better Hamiltonian support)

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -503,6 +503,10 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
 
 <h3>Bug fixes</h3>
 
+* Fixes a bug where multiple identical Hamiltonian terms will produce a
+  different result with ``optimize=True`` using ``ExpvalCost``.
+  [(#1405)](https://github.com/XanaduAI/pennylane/pull/1405)
+
```
[the remark this answers is not in the document]
Comment (antalszava) [reply .github/CHANGELOG.md:509]: Agreed, though `ExpvalCost` is user facing, so it made more sense to refer to that in the changelog. The PR should help with the details.

**Comment by antalszava** [comment]
I would expect that removing identical terms will be mean similar work to what simplifying does. So in that sense we'd have to iterate the terms and sum the identical terms.

**Pull request merged**